### PR TITLE
Add core address claiming logic

### PR DESCRIPTION
**Pull request by ad3154**
* Added the basic address claiming logic.
* Adds stub for CANMessage
* Adds stubs for some network manger functionality
* Needs still needs to be attached to the driver so it won't really do anything yet

Questions I have:

* Is there a more idomatic way to do a finite state machine than `match` ?
* Is there a setting we can adjust so the formatter doesn't make the entire state machine an unreadable blob of newlines?

### CODE REVIEW

File: src/network_management/can_message.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+pub struct CANMessage {
+    data: Vec<u8>,
+    identifier: CanId,
+    source_name: NAME,
+    destination_name: NAME,
+}
+
+impl CANMessage {
+    pub(super) fn new(data: Vec<u8>, identifier: CanId) -> CANMessage {
```
Comment (GwnDaan) [inline src/network_management/can_message.rs:16]: question: I can't really find good documentation on `pub(super)`, why is it preferred over just `pub`in this case? Is it that only this crate can create the `new` messages?

File: src/network_management/can_message.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        CANMessage {
+            data,
+            identifier,
+            source_name: NAME {
+                raw_name: DEFAULT_NAME,
+            },
+            destination_name: NAME {
+                raw_name: DEFAULT_NAME,
+            },
```
Comment (GwnDaan) [inline src/network_management/can_message.rs:25]: Maybe we should make the `DEFAULT_NAME` an instance of `NAME`? i.e. `NAME.default()` or `NAME.none()` or something

File: src/network_management/can_message.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2023 Raven Industries inc.
+#![allow(dead_code)]
```
Comment (GwnDaan) [inline src/network_management/can_message.rs:2]: opinion: I'd rather merge a failed pipeline that has these warnings than allowing them here, might be forgotten in the future otherwise

File: src/network_management/network_manager.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
+// Copyright 2023 Raven Industries inc.
+#![allow(dead_code)]
+#![allow(unused_variables)]
```
Comment (GwnDaan) [inline src/network_management/network_manager.rs:3]: Same here for these, I'd rather have the warnings in the pipeline than ignore them indefinitely

File: src/network_management/network_manager.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+use crate::network_management::can_message::CANMessage;
+use crate::network_management::common_parameter_group_numbers::CommonParameterGroupNumbers;
+use crate::network_management::name::DEFAULT_NAME;
+use crate::network_management::name::NAME;
+use std::collections::VecDeque;
+
+pub struct NetworkManager {
+    control_function_table: [Option<ControlFunction>; 253],
+    channel: u8,
+    control_function_handle_counter: usize,
```
Comment (GwnDaan) [inline src/network_management/network_manager.rs:17]: question: what will the use-case of `control_function_handle_counter` be?

File: src/network_management/network_manager.rs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    inactive_control_functions: Vec<ControlFunction>,
+    address_claim_state_machines: Vec<ControlFunction>,
```
Comment (GwnDaan) [inline src/network_management/network_manager.rs:19]: question: Is there a reason for putting the address claimers here? And not letting it be part of an 'internal' control function? Or am I too early here and is that for the future hahah

**Review by GwnDaan — COMMENTED** [review]
Looks promising 🔥 , for your questions:

> - Is there a more idomatic way to do a finite state machine than match ?

I googled around a bit, and like [this approach](https://blog.yoshuawuyts.com/state-machines-2/) quite a bit. It should also help with the next question.

> - Is there a setting we can adjust so the formatter doesn't make the entire state machine an unreadable blob of newlines?

Hmm, In my opinion refactoring the state machine into smaller concrete functions (for example as the state machine linked above, or by using early returns 😛) should be the first priority to battle this issue. Though I agree, it sometimes is unavoidable to have the longer line lengths and it may be increased just slightly (if there is an easy and standardized way to do so)

File: src/network_management/can_message.rs (hunk 1 above)
Comment (ad3154) [reply src/network_management/can_message.rs:16]: Yeah, I basically wanted to restrict use of this to only happen in `network_management` so that all message packetizing happens in there. `pub(crate)` seemed a bit too permissive, and `pub` makes it wide open basically. The idea is to have some `send_can_message` function that will eventually cause construction of a message.

File: src/network_management/can_message.rs (hunk 2 above)
Comment (ad3154) [reply src/network_management/can_message.rs:25]: `NAME.default` sounds pretty nice? I can look into that

File: src/network_management/network_manager.rs (hunk 5 above)
Comment (ad3154) [reply src/network_management/network_manager.rs:17]: Well, since we can't really do the same thing as the C++ stack when sending or receiving messages where we give the user a pointer to the control function, or in the case of `PartneredControlFunction` allow them to construct them, I figured it might make more sense to give the user a handle instead that they can then use to get the info they want out of the network manager and to prevent us having to move or copy around a lot of control function data. This is not yet done, but the idea was that the counter would be used to generate unique handles.

File: src/network_management/network_manager.rs (hunk 6 above)
Comment (ad3154) [reply src/network_management/network_manager.rs:19]: Yeah, I spent a lot of time thinking about that particular issue. Long answer... In the C++ stack we can easily keep the address claim stuff tightly bound to the ICF, which is nice. Part of what allows us to do that efficiently is our list of ICFs, which allows us to iterate over _only_ the ICFs when needed. Here you see the issue in Rust though - in the C++ world we have 2 or even 3 references to the same CF object across different lists/tables, but here in the Rust stack if I wanted to do that, since I could only have 1 owner at a time of the ICF data, the only way to reliably iterate over all ICFs would be to scan the entire address table and the entire inactive list to locate the 1 or 2 ICFs to call `update` on by matching their variant. This seemed really inefficient, and since the network manager is channel specific now, we have a guarantee that ISONAMEs are unique, so I figured there's a lot of performance to be gained by storing address claim data on its own since it will be iterated on frequently since I can still associate each address claim object back to the ICF by ISONAME if needed, which will be less frequent.

Hopefully that makes some sense? It's just to avoid scanning the whole CF table frequently.

**Comment by ad3154** [comment]
> I googled around a bit, and like [this approach](https://blog.yoshuawuyts.com/state-machines-2/) quite a bit. 

Oh, hey, that looks pretty nice... could tidy it up a fair bit. I'll experiment with that

File: src/network_management/network_manager.rs (hunk 5 above)
Comment (GwnDaan) [reply src/network_management/network_manager.rs:17]: Hmm, from what I understand it should still be doable what we do in the C++ stack, maybe in an even safer way. The smart pointers part of the official rust book should be a nice read for this:  https://doc.rust-lang.org/book/ch15-00-smart-pointers.html

Comment (ad3154) [reply src/network_management/network_manager.rs:17]: Hmm... if we went to `Rc` or `Arc` I might also then be able to represent address claim data similarly to how we did in the C++ stack, as I could clone `Rc`s into their own list for new ICFs...

File: src/network_management/can_message.rs (hunk 3 above)
Comment (Notgnoshi) [reply src/network_management/can_message.rs:2]: > might be forgotten in the future otherwise

This has definitely happened to me in other projects. `_` prefixes, and `#![allow]` macros are hard to remove once they're present.

File: src/network_management/network_manager.rs (hunk 5 above)
Comment (GwnDaan) [reply src/network_management/network_manager.rs:17]: Yeah indeed, I think it can end up a lot cleaner with having the different CF types. I expect this network_manager will otherwise grow _really_ quickly otherwise, since we can not use global statics anymore unlike C++ and almost everything will come together here. So probably best to avoid putting complex things here when feasible

File: src/network_management/network_manager.rs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+    fn send_raw_can_message(
+        &self,
+        source_address: u8,
+        destination_address: u8,
+        parameter_group_number: u32,
+        priority: u8,
+        data: &[u8],
+    ) -> bool {
+        false // Todo, link up with the driver layer
```
Comment (Notgnoshi) [inline src/network_management/network_manager.rs:49]: offtopic: Rust provides a few macros that are useful for stuff like this, depending on whether you want to be able to call this function without crashing or not.

* `unreachable!("Reason this branch is unreachable")`
* `unimplemented!("This feature isn't implemented yet")`
* `todo!("This should be handled")`

I'm not 100% sure how the intent is different between `todo!` and `unimplemented!`, but they're pretty nice to use, because they can change how the code is compiled.

```rust
enum Strategy {
  Foo,
  Bar,
  Baz,
}

let result = match strategy {
  Strategy::Foo => 2 + 2,
  Strategy::Bar => 0xdeadbeef,
  Strategy::Qux => unimplemented!("The Qux strategy isn't implemented yet"),
};
```
If `strategy` were `Qux`, this code would crash with the given error message, but notice the code compiles as if `unimplmented!` returned the right value of the right type.

I don't often reach for these macros, because they result in a crash, and I often want to use a dummy placeholder value instead so I can work through plumbing first.

File: src/network_management/network_manager.rs [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        let address_claim = vec![
+            (name & 0xFF) as u8,
+            ((name >> 8) & 0xFF) as u8,
+            ((name >> 16) & 0xFF) as u8,
+            ((name >> 24) & 0xFF) as u8,
+            ((name >> 32) & 0xFF) as u8,
+            ((name >> 40) & 0xFF) as u8,
+            ((name >> 48) & 0xFF) as u8,
+            ((name >> 56) & 0xFF) as u8,
+        ];
```
Comment (Notgnoshi) [inline src/network_management/network_manager.rs:76]: I'm not 100% sure this would work, but there's also https://doc.rust-lang.org/std/primitive.u64.html#method.to_le_bytes (with various endiannesses)

File: src/network_management/can_message.rs (hunk 2 above)
Comment (Notgnoshi) [reply src/network_management/can_message.rs:25]: There's two approaches I'd suggest

```rust
impl NAME {
    pub const DEFAULT: NAME = NAME::new(0xFFFFF);
}
```
or
```rust
impl Default for NAME {
    fn default() -> NAME {
        NAME::new(0xFFFFFFFFFFF)
    }
}
```
They have different strengths. The second one allows

```rust
#[derive(Default)]  // <--- this only works if all of the types contained in the struct also impl Default, oitherwise you have to implement it manually
struct Foo {
    name: NAME,
    x: f64,
    y: f64,
}
```
but the first one is maybe more natural if there are other associated constants (like there are for `Address` or `Pgn`)

**Comment by ad3154** [comment]
> > I googled around a bit, and like [this approach](https://blog.yoshuawuyts.com/state-machines-2/) quite a bit.
> 
> Oh, hey, that looks pretty nice... could tidy it up a fair bit. I'll experiment with that

Am I missing something? That approach doesn't seem to compile....
image

**Comment by GwnDaan** [comment]
> > > I googled around a bit, and like [this approach](https://blog.yoshuawuyts.com/state-machines-2/) quite a bit.
> > 
> > 
> > Oh, hey, that looks pretty nice... could tidy it up a fair bit. I'll experiment with that
> 
> Am I missing something? That approach doesn't seem to compile.... image

Welp I missed the most important part of the post:
>In this post I instead want to focus on ✨ the future ✨. How could we plausibly represent this in Rust by extending the language. In the "future directions" section of the last post we wrote a state machine without the blinky lights. Using [enum variants types](https://github.com/rust-lang/rfcs/pull/1450) (not yet a thing) combined with [arbitrary self-types](https://github.com/rust-lang/rust/issues/44874) (also not yet a thing), we could write the following:

So that suggestion doesn't work hahah, sorry

**Comment by ad3154** [comment]
> So that suggestion doesn't work hahah, sorry

Bummer, it was so clean and nice! Regardless, I'll still try and make it more self-contained by moving logic into `AddressClaimingState`.

File: src/network_management/network_manager.rs (hunk 5 above)
Comment (ad3154) [reply src/network_management/network_manager.rs:17]: One wrinkle I'm running into is that `Rc` only provides immutable access, so although we could use them to pass CFs around as you described, I can't use it to shortcut my other issue where I wanted to iterate over a shorter list of references to CF objects and mutate them... so I am back to having to either having to scan the whole table every few milliseconds, or store address claiming data in the network manager 😣

Comment (GwnDaan) [reply src/network_management/network_manager.rs:17]: Owh noo :( Maybe chapter 15.5 (smart pointers, interior mutability) will help you out hahah: https://doc.rust-lang.org/book/ch15-05-interior-mutability.html. In particular from the section [Having Multiple Owners of Mutable Data by Combining Rc<T> and RefCell<T>](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html#having-multiple-owners-of-mutable-data-by-combining-rct-and-refcellt) is interesting

TLDR: try to use `Rc<RefCell<...>>`

It's a long time since I used it but it should do the job here IIRC

**Comment by Thom-de-Jong** [comment]
Just want to add to this conversation.

State machines can be modeled in Rust using using enum's, as one would normally do in c++.
State machines however, really shine when combined with Rust's compile-time guarantees, making it impossible to program a transition to a state that we should not be in. ([Typestate Programming](https://docs.rust-embedded.org/book/static-guarantees/typestate-programming.html#typestate-programming))

The best example, that i know of, explaining this concept for state machines can be found here ([Pretty State Machine Patterns in Rust ](https://hoverbear.org/blog/rust-state-machine-pattern/))

Note however that this relatively requires a lot of code. Luckily most of it will be abstracted away by the compiler.
I would say, if we use Rust for its awesome, safe, compiler, we should build our software so the compiler will check for invalid states.

What do you think?

**Comment by GwnDaan** [comment]
@Thom-de-Jong that sounds interesting! I think we should at least try it, though I'm also fine with doing at a later stage with a seperate PR if that's easier for @ad3154

> Note however that this relatively requires a lot of code.

Personally that is fine with me, the large switch/match statements for some of the state machines we have over in AgIsoStack++ are hard to gasp IMO. So anything resulting is a somewhat smaller functions of code and especially with the "Rust's compile-time guarantees, making it impossible to program a transition to a state that we should not be in." I'm in for.

**Comment by ad3154** [comment]
Hmm, that does seem nice, but agreed it's a lot of code. I'd be fine with trying it but yeah maybe I can follow up this PR with a refactor. We have a _ton_ of state machines in AgIsoStack++ so getting a nice pattern figured out will make our lives easier for other parts of the stack as well, so thanks for the suggestion!

File: src/network_management/can_message.rs (hunk 2 above)
Comment (ad3154) [reply src/network_management/can_message.rs:22]: Waiting on #17  or #9 to convert to this

**Comment by ad3154** [comment]
I think this is in a good enough state to merge now so we can keep making progress. Since last time, I rebased and updated all usages of NAME in each commit to match the latest implementation of NAME. Would be nice to get this in before #26, and will give us something I can test the driver layer with

**Review by GwnDaan — APPROVED** [review]
